// @textile/powergate-client/dist/ffs/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __asyncValues = (this && this.__asyncValues) || function (o) {
    if (!Symbol.asyncIterator) throw new TypeError("Symbol.asyncIterator is not defined.");
    var m = o[Symbol.asyncIterator], i;
    return m ? m.call(o) : (o = typeof __values === "function" ? __values(o) : o[Symbol.iterator](), i = {}, verb("next"), verb("throw"), verb("return"), i[Symbol.asyncIterator] = function () { return this; }, i);
    function verb(n) { i[n] = o[n] && function (v) { return new Promise(function (resolve, reject) { v = o[n](v), settle(resolve, reject, v.done, v.value); }); }; }
    function settle(resolve, reject, d, v) { Promise.resolve(v).then(function(v) { resolve({ value: v, done: d }); }, reject); }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createFFS = void 0;
const grpc_web_1 = require("@improbable-eng/grpc-web");
const rpc_pb_1 = require("@textile/grpc-powergate-client/dist/ffs/rpc/rpc_pb");
const rpc_pb_service_1 = require("@textile/grpc-powergate-client/dist/ffs/rpc/rpc_pb_service");
const fs_1 = __importDefault(require("fs"));
const ipfs_http_client_1 = __importDefault(require("ipfs-http-client"));
const it_block_1 = __importDefault(require("it-block"));
const path_1 = __importDefault(require("path"));
const util_1 = require("../util");
const normalize_1 = require("./normalize");
const util_2 = require("./util");
/**
 * @ignore
 */
exports.createFFS = (config, getMeta, getHeaders) => {
    const client = new rpc_pb_service_1.RPCServiceClient(config.host, config);
    const ipfs = ipfs_http_client_1.default(config.host);
    return {
        create: () => util_1.promise((cb) => client.create(new rpc_pb_1.CreateRequest(), cb), (res) => res.toObject()),
        list: () => util_1.promise((cb) => client.listAPI(new rpc_pb_1.ListAPIRequest(), cb), (res) => res.toObject()),
        id: () => util_1.promise((cb) => client.iD(new rpc_pb_1.IDRequest(), getMeta(), cb), (res) => res.toObject()),
        addrs: () => util_1.promise((cb) => client.addrs(new rpc_pb_1.AddrsRequest(), getMeta(), cb), (res) => res.toObject()),
        defaultStorageConfig: () => util_1.promise((cb) => client.defaultStorageConfig(new rpc_pb_1.DefaultStorageConfigRequest(), getMeta(), cb), (res) => res.toObject()),
        newAddr: (name, type, makeDefault) => {
            const req = new rpc_pb_1.NewAddrRequest();
            req.setName(name);
            req.setAddressType(type || "bls");
            req.setMakeDefault(makeDefault || false);
            return util_1.promise((cb) => client.newAddr(req, getMeta(), cb), (res) => res.toObject());
        },
        getStorageConfig: (cid) => {
            const req = new rpc_pb_1.GetStorageConfigRequest();
            req.setCid(cid);
            return util_1.promise((cb) => client.getStorageConfig(req, getMeta(), cb), (res) => res.toObject());
        },
        setDefaultStorageConfig: (config) => {
            const c = new rpc_pb_1.StorageConfig();
            c.setRepairable(config.repairable);
            if (config.hot) {
                c.setHot(util_2.hotObjToMessage(config.hot));
            }
            if (config.cold) {
                c.setCold(util_2.coldObjToMessage(config.cold));
            }
            const req = new rpc_pb_1.SetDefaultStorageConfigRequest();
            req.setConfig(c);
            return util_1.promise((cb) => client.setDefaultStorageConfig(req, getMeta(), cb), () => {
                // nothing to return
            });
        },
        show: (cid) => {
            const req = new rpc_pb_1.ShowRequest();
            req.setCid(cid);
            return util_1.promise((cb) => client.show(req, getMeta(), cb), (res) => res.toObject());
        },
        info: () => util_1.promise((cb) => client.info(new rpc_pb_1.InfoRequest(), getMeta(), cb), (res) => res.toObject()),
        getStorageJob: (jobId) => {
            const req = new rpc_pb_1.GetStorageJobRequest();
            req.setJid(jobId);
            return util_1.promise((cb) => client.getStorageJob(req, getMeta(), cb), (res) => res.toObject());
        },
        watchJobs: (handler, ...jobs) => {
            const req = new rpc_pb_1.WatchJobsRequest();
            req.setJidsList(jobs);
            const stream = client.watchJobs(req, getMeta());
            stream.on("data", (res) => {
                var _a;
                const job = (_a = res.getJob()) === null || _a === void 0 ? void 0 : _a.toObject();
                if (job) {
                    handler(job);
                }
            });
            return stream.cancel;
        },
        cancelJob: (jobId) => {
            const req = new rpc_pb_1.CancelJobRequest();
            req.setJid(jobId);
            return util_1.promise((cb) => client.cancelJob(req, getMeta(), cb), () => {
                // nothing to return
            });
        },
        watchLogs: (handler, cid, opts = {}) => {
            const req = new rpc_pb_1.WatchLogsRequest();
            req.setCid(cid);
            if (opts.includeHistory) {
                req.setHistory(opts.includeHistory);
            }
            if (opts.jobId) {
                req.setJid(opts.jobId);
            }
            const stream = client.watchLogs(req, getMeta());
            stream.on("data", (res) => {
                var _a;
                const logEntry = (_a = res.getLogEntry()) === null || _a === void 0 ? void 0 : _a.toObject();
                if (logEntry) {
                    handler(logEntry);
                }
            });
            return stream.cancel;
        },
        replace: (cid1, cid2) => {
            const req = new rpc_pb_1.ReplaceRequest();
            req.setCid1(cid1);
            req.setCid2(cid2);
            return util_1.promise((cb) => client.replace(req, getMeta(), cb), (res) => res.toObject());
        },
        pushStorageConfig: (cid, opts = {}) => {
            const req = new rpc_pb_1.PushStorageConfigRequest();
            req.setCid(cid);
            if (opts.override) {
                req.setOverrideConfig(opts.override);
                req.setHasOverrideConfig(true);
            }
            if (opts.storageConfig) {
                const c = new rpc_pb_1.StorageConfig();
                c.setRepairable(opts.storageConfig.repairable);
                if (opts.storageConfig.hot) {
                    c.setHot(util_2.hotObjToMessage(opts.storageConfig.hot));
                }
                if (opts.storageConfig.cold) {
                    c.setCold(util_2.coldObjToMessage(opts.storageConfig.cold));
                }
                req.setConfig(c);
                req.setHasConfig(true);
            }
            return util_1.promise((cb) => client.pushStorageConfig(req, getMeta(), cb), (res) => res.toObject());
        },
        remove: (cid) => {
            const req = new rpc_pb_1.RemoveRequest();
            req.setCid(cid);
            return util_1.promise((cb) => client.remove(req, getMeta(), cb), () => {
                // nothing to return
            });
        },
        get: (cid) => {
            return new Promise((resolve, reject) => {
                const append = (l, r) => {
                    const tmp = new Uint8Array(l.byteLength + r.byteLength);
                    tmp.set(l, 0);
                    tmp.set(r, l.byteLength);
                    return tmp;
                };
                let final = new Uint8Array();
                const req = new rpc_pb_1.GetRequest();
                req.setCid(cid);
                const stream = client.get(req, getMeta());
                stream.on("data", (resp) => {
                    final = append(final, resp.getChunk_asU8());
                });
                stream.on("end", (status) => {
                    if ((status === null || status === void 0 ? void 0 : status.code) !== grpc_web_1.grpc.Code.OK) {
                        reject(`error code ${status === null || status === void 0 ? void 0 : status.code} - ${status === null || status === void 0 ? void 0 : status.details}`);
                    }
                    else {
                        resolve(final);
                    }
                });
            });
        },
        getFolder: (cid, output, opts = {}) => __awaiter(void 0, void 0, void 0, function* () {
            var e_1, _a, e_2, _b;
            const headers = getHeaders();
            const options = { headers };
            if (opts.timeout) {
                options["timeout"] = opts.timeout;
            }
            try {
                for (var _c = __asyncValues(ipfs.get(cid, options)), _d; _d = yield _c.next(), !_d.done;) {
                    const file = _d.value;
                    const noCidPath = file.path.replace(cid, "");
                    const fullFilePath = path_1.default.join(output, noCidPath);
                    if (file.content) {
                        yield fs_1.default.promises.mkdir(path_1.default.join(output, path_1.default.dirname(file.path)), { recursive: true });
                        const stream = fs_1.default.createWriteStream(fullFilePath);
                        try {
                            for (var _e = (e_2 = void 0, __asyncValues(file.content)), _f; _f = yield _e.next(), !_f.done;) {
                                const chunk = _f.value;
                                const slice = chunk.slice();
                                yield new Promise((resolve, reject) => {
                                    stream.write(slice, (err) => {
                                        if (err) {
                                            reject(err);
                                        }
                                        else {
                                            resolve();
                                        }
                                    });
                                });
                            }
                        }
                        catch (e_2_1) { e_2 = { error: e_2_1 }; }
                        finally {
                            try {
                                if (_f && !_f.done && (_b = _e.return)) yield _b.call(_e);
                            }
                            finally { if (e_2) throw e_2.error; }
                        }
                    }
                    else {
                        // this is a dir
                        yield fs_1.default.promises.mkdir(fullFilePath, { recursive: true });
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_d && !_d.done && (_a = _c.return)) yield _a.call(_c);
                }
                finally { if (e_1) throw e_1.error; }
            }
        }),
        sendFil: (from, to, amount) => {
            const req = new rpc_pb_1.SendFilRequest();
            req.setFrom(from);
            req.setTo(to);
            req.setAmount(amount);
            return util_1.promise((cb) => client.sendFil(req, getMeta(), cb), () => {
                // nothing to return
            });
        },
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        stage: (input) => __awaiter(void 0, void 0, void 0, function* () {
            // Only process the first  input if there are more than one
            const source = (yield normalize_1.normaliseInput(input).next()).value;
            return new Promise((resolve, reject) => __awaiter(void 0, void 0, void 0, function* () {
                var e_3, _g;
                const client = grpc_web_1.grpc.client(rpc_pb_service_1.RPCService.Stage, config);
                client.onMessage((message) => {
                    resolve(message.toObject());
                });
                client.onEnd((code, msg) => {
                    if (code !== grpc_web_1.grpc.Code.OK) {
                        reject(`error code ${code} - ${msg}`);
                    }
                    else {
                        reject("ended with no message");
                    }
                });
                if (source === null || source === void 0 ? void 0 : source.content) {
                    client.start(getMeta());
                    const process = yield it_block_1.default({ size: 32000, noPad: true });
                    try {
                        for (var _h = __asyncValues(process(source.content)), _j; _j = yield _h.next(), !_j.done;) {
                            const chunk = _j.value;
                            const buf = chunk.slice();
                            const req = new rpc_pb_1.StageRequest();
                            req.setChunk(buf);
                            client.send(req);
                        }
                    }
                    catch (e_3_1) { e_3 = { error: e_3_1 }; }
                    finally {
                        try {
                            if (_j && !_j.done && (_g = _h.return)) yield _g.call(_h);
                        }
                        finally { if (e_3) throw e_3.error; }
                    }
                    client.finishSend();
                }
                else {
                    reject(new Error("no content to stage"));
                }
            }));
        }),
        stageFolder: (path) => __awaiter(void 0, void 0, void 0, function* () {
            const src = ipfs_http_client_1.default.globSource(path, { recursive: true });
            const headers = getHeaders();
            const res = yield ipfs.add(src, { headers });
            return res.cid.string;
        }),
        listPayChannels: () => util_1.promise((cb) => client.listPayChannels(new rpc_pb_1.ListPayChannelsRequest(), getMeta(), cb), (res) => res.toObject()),
        createPayChannel: (from, to, amt) => {
            const req = new rpc_pb_1.CreatePayChannelRequest();
            req.setFrom(from);
            req.setTo(to);
            req.setAmount(amt);
            return util_1.promise((cb) => client.createPayChannel(req, getMeta(), cb), (res) => res.toObject());
        },
        redeemPayChannel: (payChannelAddr) => {
            const req = new rpc_pb_1.RedeemPayChannelRequest();
            req.setPayChannelAddr(payChannelAddr);
            return util_1.promise((cb) => client.redeemPayChannel(req, getMeta(), cb), () => {
                // nothing to return
            });
        },
        listStorageDealRecords: (opts = {}) => {
            const req = new rpc_pb_1.ListStorageDealRecordsRequest();
            req.setConfig(listDealRecordsOptionsToConfig(opts));
            return util_1.promise((cb) => client.listStorageDealRecords(req, getMeta(), cb), (res) => res.toObject());
        },
        listRetrievalDealRecords: (opts = {}) => {
            const req = new rpc_pb_1.ListRetrievalDealRecordsRequest();
            req.setConfig(listDealRecordsOptionsToConfig(opts));
            return util_1.promise((cb) => client.listRetrievalDealRecords(req, getMeta(), cb), (res) => res.toObject());
        },
        showAll: () => util_1.promise((cb) => client.showAll(new rpc_pb_1.ShowAllRequest(), getMeta(), cb), (res) => res.toObject()),
    };
};
function listDealRecordsOptionsToConfig(opts) {
    const conf = new rpc_pb_1.ListDealRecordsConfig();
    if (opts.ascending) {
        conf.setAscending(opts.ascending);
    }
    if (opts.dataCids) {
        conf.setDataCidsList(opts.dataCids);
    }
    if (opts.fromAddresses) {
        conf.setFromAddrsList(opts.fromAddresses);
    }
    if (opts.includeFinal) {
        conf.setIncludeFinal(opts.includeFinal);
    }
    if (opts.includePending) {
        conf.setIncludePending(opts.includePending);
    }
    return conf;
}
